src/Libraries/Microsoft.Extensions.Http.Telemetry/Logging/Internal/HttpClientLogger.cs

```diff
@@ -0,0 +1,234 @@
+
+        _logRequestStart = optionsValue.LogRequestStart;
+        _logResponseHeaders = optionsValue.ResponseHeadersDataClasses.Count > 0;
+        _logRequestHeaders = optionsValue.RequestHeadersDataClasses.Count > 0;
+        _pathParametersRedactionSkipped = optionsValue.RequestPathParameterRedactionMode == HttpRouteParameterRedactionMode.None;
+    }
+
+    [SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "The logger shouldn't throw")]
+    public async ValueTask<object?> LogRequestStartAsync(HttpRequestMessage request, CancellationToken cancellationToken = default)
+    {
+        var logRecord = _logRecordPool.Get();
+
+        List<KeyValuePair<string, string>>? requestHeadersBuffer = null;
+        if (_logRequestHeaders)
+        {
+            requestHeadersBuffer = _headersPool.Get();
+        }
+
+        try
+        {
+            await _httpRequestReader.ReadRequestAsync(logRecord, request, requestHeadersBuffer, cancellationToken).ConfigureAwait(false);
+
+            if (_logRequestStart)
+            {
+                Log.OutgoingRequest(_logger, LogLevel.Information, logRecord);
+            }
+
+            return logRecord;
+        }
+        catch (Exception ex)
+        {
+            // If redaction is skipped, we can log unredacted request path; otherwise use "logRecord.Path" (even though it might not be set):
+            var pathToLog = _pathParametersRedactionSkipped
+                ? request.RequestUri?.AbsolutePath
+                : logRecord.Path;
+
+            Log.RequestReadError(_logger, ex, request.Method, request.RequestUri?.Host, pathToLog ?? string.Empty);
+
+            // Return back pooled objects (since the logRecord wasn't fully prepared):
+            _logRecordPool.Return(logRecord);
+
+            if (requestHeadersBuffer is not null)
+            {
+                _headersPool.Return(requestHeadersBuffer);
+            }
+
+            // Recommendation is to swallow the exception (logger shouldn't throw), so we don't re-throw here:
+            return null;
+        }
+    }
+
+    public async ValueTask LogRequestStopAsync(
+        object? context,
+        HttpRequestMessage request,
+        HttpResponseMessage response,
+        TimeSpan elapsed,
+        CancellationToken cancellationToken = default)
+            => await LogResponseAsync(context, request, response, null, elapsed, cancellationToken).ConfigureAwait(false);
+
+    public async ValueTask LogRequestFailedAsync(
+        object? context,
+        HttpRequestMessage request,
+        HttpResponseMessage? response,
+        Exception exception,
+        TimeSpan elapsed,
+        CancellationToken cancellationToken = default)
+            => await LogResponseAsync(context, request, response, exception, elapsed, cancellationToken).ConfigureAwait(false);
+
+    public object? LogRequestStart(HttpRequestMessage request)
+    {
+        throw new NotSupportedException();
```

Comment:
As of now the best we can offer for sync overloads - is to call their async counterparts with `.GetAwaiter().GetResult()`, that's because there's no sync API of reading request/response message body